### PR TITLE
Implement a few methods for RingBuf

### PR DESCRIPTION
**Pull request by pczarn**
Part of collections reform part 1 and 2, #18424 and #19986
- shrink_to_fit
- swap_back_remove
- swap_front_remove
- truncate
- resize

### CODE REVIEW

**Comment by Gankro** [comment]
@pczarn I see you've added some commits; is that everything or are you still working on it (basically: should I re-review now?)

**Comment by pczarn** [comment]
@Gankro that's everything, except for swap_.._remove tests; hope to add them now.

File: src/libcollections/ring_buf.rs [hunk 1]
```diff
@@ -121,6 +118,19 @@ impl<T> RingBuf<T> {
...
+    /// Copies a contiguous block of memory len long from src to dst
+    #[inline]
+    unsafe fn copy_nonoverlapping(&self, dst: uint, src: uint, len: uint) {
+        debug_assert!(dst + len <= self.cap, "dst={} src={} len={} cap={}", dst, src, len,
+                      self.cap);
+        debug_assert!(src + len <= self.cap, "dst={} src={} len={} cap={}", dst, src, len,
+                      self.cap);
+        ptr::copy_nonoverlapping_memory(
+            self.ptr.offset(dst as int),
+            self.ptr.offset(src as int) as *const T,
```
Comment (Gankro) [inline src/libcollections/ring_buf.rs]: I *think* the `as *const T` should just coerce for free? Still trying to grok when these coercions do or don't kick in, though.

**Comment by Gankro** [comment]
Okay lgtm modulo nits/tests.

**Comment by Gankro** [comment]
r=me with squash

**Comment by pczarn** [comment]
@Gankro done.

**Pull request merged**